### PR TITLE
rgw: ability to list bucket contents in unsorted order for efficiency

### PR DESCRIPTION
**Pull request by ivancich**
rgw: ability to list bucket contents in unsorted order for efficiency

Add the ability to list the contents of a bucket without the results
being sorted. This can have performance benefits since all bucket
shards do not need to be queried. This is done by adding a parameter
to the REST query string (i.e., "allow_unordered" [note underscore]
for swift or "allow-unordered" [note hyphen] for s3) to true.

This works with other request parameters such as marker, end_marker,
and prefix. But it does not work when delimeter is used. Also, because
the results are not in order, a subsequent marker may precede an
earlier marker.

Signed-off-by: J. Eric Ivancich <ivancich@redhat.com>

### CODE REVIEW

**Comment by ivancich** [comment]
Note 1: I have a PR for the swift tests that cannot be merged until this is merged, as tests will fail without this rgw modification.

Note 2: I have added tests for @alimaredia added boto3 tests, which have been merged into his branch and will ultimately be merged into the master of s3-tests.

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -5772,7 +5782,140 @@ int RGWRados::Bucket::List::list_objects(int64_t max,
...
+  int shard_id = target->get_shard_id();
+
+  int count = 0;
+  bool truncated = true;
+  int read_ahead;
+  if (false) {
+    read_ahead = std::max(cct->_conf->rgw_list_bucket_min_readahead, max);
+  } else {
+    read_ahead = 2 * max;
+  }
```
[the remark this answers is not in the document]
Comment (ivancich) [reply src/rgw/rgw_rados.cc]: I don't believe there's any benefit to reading a minimum of 1,000 when caller requests a much smaller value. Should read some extra due to filtering stages below.

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -12892,25 +13048,36 @@ int RGWRados::cls_obj_set_bucket_tag_timeout(RGWBucketInfo& bucket_info, uint64_
...
+				      const string& prefix,
+				      uint32_t num_entries,
+				      bool list_versions,
+				      map<string, rgw_bucket_dir_entry>& m,
+				      bool *is_truncated,
+				      rgw_obj_index_key *last_entry,
+				      bool (*force_check_filter)(const string& name))
 {
-  ldout(cct, 10) << "cls_bucket_list " << bucket_info.bucket << " start " << start.name << "[" << start.instance << "] num_entries " << num_entries << dendl;
+  ldout(cct, 10) << "cls_bucket_list " << bucket_info.bucket << " start " <<
```
Comment (ZVampirEM77) [inline src/rgw/rgw_rados.cc]: cls_bucket_list_ordered is better, right?

File: doc/radosgw/s3/bucketops.rst [hunk 3]
```diff
@@ -91,17 +91,19 @@ Syntax
...
 +-----------------+-----------+-----------------------------------------------------------------------+
-| ``delimiter``   | String    | The delimiter between the prefix and the rest of the object name.     |
+| ``marker``          | String    | A beginning index for the list of objects returned.                                            |
 +-----------------+-----------+-----------------------------------------------------------------------+
-| ``marker``      | String    | A beginning index for the list of objects returned.                   |
-+-----------------+-----------+-----------------------------------------------------------------------+
-| ``max-keys``    | Integer   | The maximum number of keys to return. Default is 1000.                |
+| ``max-keys``        | Integer   | The maximum number of keys to return. Default is 1000.                                         |
 +-----------------+-----------+-----------------------------------------------------------------------+
+| ``allow-unordered`` | Boolean   | Non-standard extension. Allows reults to be returned unordered. Cannot be used with delimiter. |
```
Comment (cbodley) [inline doc/radosgw/s3/bucketops.rst]: reults ^

File: src/rgw/rgw_rados.h [hunk 4]
```diff
@@ -3040,19 +3053,35 @@ class RGWRados : public AdminSocketHook
...
+	  allow_unordered(false)
+	{}
       } params;
 
-    public:
       explicit List(RGWRados::Bucket *_target) : target(_target) {}
 
-      int list_objects(int64_t max, vector<rgw_bucket_dir_entry> *result, map<string, bool> *common_prefixes, bool *is_truncated);
+      int list_objects(int64_t max,
+		       vector<rgw_bucket_dir_entry> *result,
```
[the remark this answers is not in the document]
Comment (ivancich) [reply src/rgw/rgw_rados.h]: Great idea! I made the change from list to vector.

File: src/rgw/rgw_rados.cc (hunk 2 above)
Comment (ivancich) [reply src/rgw/rgw_rados.cc]: Yes! And the logging output later in the function as well!

**Comment by ivancich** [comment]
jenkins test docs

**Comment by ivancich** [comment]
jenkins test docs

**Comment by cbodley** [comment]
jenkins render docs

File: src/cls/rgw/cls_rgw_ops.h [hunk 5]
```diff
@@ -413,8 +413,7 @@ struct rgw_cls_list_op
 };
 WRITE_CLASS_ENCODER(rgw_cls_list_op)
 
-struct rgw_cls_list_ret
-{
+struct rgw_cls_list_ret {
```
Comment (mattbenjamin) [inline src/cls/rgw/cls_rgw_ops.h]: I don't care, but is this preferred?

Comment (ivancich) [reply src/cls/rgw/cls_rgw_ops.h]: https://github.com/ceph/ceph/blob/master/CodingStyle cites https://google.github.io/styleguide/cppguide.html for C++ style. And in all their examples of class definitions, function definitions, and control structures, the opening brace is on the end of the line, not on a line of its own.

Higher up in this source file both styles are used for struct declarations. As I go through the code trying to understand I'll sometimes try to make things a bit more consistent and add comments that make it easier for me. (And I have my own biases.)

**Comment by ivancich** [comment]
jenkins test make check

**Comment by ivancich** [comment]
jenkins test make check arm64

File: doc/radosgw/s3/bucketops.rst [hunk 6]
```diff
@@ -91,17 +91,19 @@ Syntax
...
+| ``delimiter``       | String    | The delimiter between the prefix and the rest of the object name.                               |
++-----------------+-----------+-----------------------------------------------------------------------------------------------------+
+| ``marker``          | String    | A beginning index for the list of objects returned.                                             |
++-----------------+-----------+-----------------------------------------------------------------------------------------------------+
+| ``max-keys``        | Integer   | The maximum number of keys to return. Default is 1000.                                          |
++-----------------+-----------+-----------------------------------------------------------------------------------------------------+
+| ``allow-unordered`` | Boolean   | Non-standard extension. Allows results to be returned unordered. Cannot be used with delimiter. |
++-----------------+-----------+-----------------------------------------------------------------------------------------------------+
```
Comment (cbodley) [inline doc/radosgw/s3/bucketops.rst]: it doesn't look like this table is rendering correctly: http://docs.ceph.com/ceph-prs/21026/radosgw/s3/bucketops/#id4

i think the `++-----------------+` parts need to be extended to match?

**Comment by ivancich** [comment]
jenkins render docs

File: doc/radosgw/s3/bucketops.rst (hunk 6 above)
Comment (ivancich) [reply doc/radosgw/s3/bucketops.rst]: Thanks, Casey. They were the right length. But the "+"s weren't aligned correctly.

**Comment by ivancich** [comment]
jenkins render docs

**Comment by ivancich** [comment]
@cbodley Rendering is now fixed. Will dig into the build issues later today....

**Comment by liewegas** [comment]
retest this please

**Comment by ivancich** [comment]
Tested this on an ARM machine from RH Beaker lab. Passed all tests.

> 100% tests passed, 0 tests failed out of 152
> 
> Total Test time (real) = 1864.22 sec
> cmake check: successful run on 6da5a5888c8605497cddc83b73cee1528d1b4b44

> [ivancich@apm-mustang-ev3-03 ceph]$ lscpu
> Architecture:        aarch64
> Byte Order:          Little Endian
> CPU(s):              8
> On-line CPU(s) list: 0-7
> Thread(s) per core:  1
> Core(s) per socket:  2
> Socket(s):           4
> NUMA node(s):        1
> Model:               0
> BogoMIPS:            100.00
> L1d cache:           unknown size
> L1i cache:           unknown size
> L2 cache:            unknown size
> NUMA node0 CPU(s):   0-7
> Flags:               fp asimd evtstrm cpuid

**Comment by ivancich** [comment]
jenkins retest this please

**Comment by tchaikov** [comment]
@cbodley and @ivancich seems this breaks a test in rados qa suite, i filed http://tracker.ceph.com/issues/23699, could you take a look at your convenience?

**Pull request merged**